nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1095 @@
+        for p in self:
+            if isinstance(p, Cifti2Volume):
+                return p
+        return None
+
+    @volume.setter
+    def volume(self, volume):
+        if not isinstance(volume, Cifti2Volume):
+            raise ValueError("You can only set a volume with a volume")
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            self.append(volume)
+            return
+        self[i] = volume
+
+    @volume.deleter
+    def volume(self):
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            raise ValueError("No Cifti2Volume element")
+        del self[i]
+
+    @property
+    def brain_models(self):
+        for p in self:
+            if isinstance(p, Cifti2BrainModel):
+                yield p
+
+    def _to_xml_element(self):
+        if self.applies_to_matrix_dimension is None:
+            raise CIFTI2HeaderError(
+                'MatrixIndicesMap element requires to be applied to at least 1 dimension'
+            )
+
+        mat_ind_map = xml.Element('MatrixIndicesMap')
+        dims_as_strings = [str(dim) for dim in self.applies_to_matrix_dimension]
+        mat_ind_map.attrib['AppliesToMatrixDimension'] = ','.join(dims_as_strings)
+        for key in ['IndicesMapToDataType', 'NumberOfSeriesPoints', 'SeriesExponent',
+                    'SeriesStart', 'SeriesStep', 'SeriesUnit']:
+            attr = _underscore(key)
+            value = getattr(self, attr)
+            if value is not None:
+                mat_ind_map.attrib[key] = str(value)
+        for map_ in self:
+            mat_ind_map.append(map_._to_xml_element())
+
+        return mat_ind_map
+
+
+class Cifti2Matrix(xml.XmlSerializable, collections.MutableSequence):
+    def __init__(self):
+        self._mims = []
+        self.metadata = None
+
+    @property
+    def metadata(self):
+        return self._meta
+
+    @metadata.setter
+    def metadata(self, meta):
+        """ Set the metadata for this Cifti2Header
+
+        Parameters
+        ----------
+        meta : Cifti2MetaData
+
+        Returns
+        -------
+        None
+        """
+        self._meta = _value_if_klass(meta, Cifti2MetaData)
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims[key] = value
+
+    def __getitem__(self, key):
+        return self._mims[key]
+
+    def __delitem__(self, key):
+        del self._mims[key]
+
+    def __len__(self):
+        return len(self._mims)
+
+    def insert(self, index, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims.insert(index, value)
+
+    def _to_xml_element(self):
+        if (len(self) == 0 and self.metadata is None):
```

Comment:
Yes, that is my major worry about this PR.   I think the original speculation was that just replicating the XML elements in Python would be a good foundation for a more user-friendly API.   That might be true, but it's very difficult to tell at the moment.   In the meantime, if we put this into a released version, we'll have to keep this API, because the current code loads the XML into these objects in the header, which is a visible API:

```
In [1]: import nibabel.cifti2 as ci                             
In [2]: img = ci.load('nibabel-data/nitest-cifti2/ones.dscalar.nii')
In [3]: img.header
Out[3]: <nibabel.cifti2.cifti2.Cifti2Header at 0x10809f2d0> 
In [4]: img.header.matrix
Out[4]: <nibabel.cifti2.cifti2.Cifti2Matrix at 0x10809f290>
```

So we have a trade-off, where we gain from being able to load CIFTI2 image in some form, possibly hard to use, and we may lose by having to support an low-level API that is an uncomfortable basis for a better one.

At the moment, we don't have much of a guess for that second cost.   One way to get at it, would be seeing what it would look like to construct a few valid headers for realistic data - for example, replicating the headers of the example data without loading them.